cpu/sam0_common/usb/usb.c

```diff
@@ -0,0 +1,679 @@
+        case USB_EP_TYPE_ISOCHRONOUS:
+            type = 0x02;
+            break;
+        case USB_EP_TYPE_BULK:
+            type = 0x03;
+            break;
+        case USB_EP_TYPE_INTERRUPT:
+            type = 0x04;
+            break;
+        case USB_EP_TYPE_NONE:
+            /* Should never happen */
+            assert(false);
+    }
+    if (ep->dir == USB_EP_DIR_OUT) {
+        ep_reg->EPCFG.bit.EPTYPE0 = type;
+    }
+    else {
+        ep_reg->EPCFG.bit.EPTYPE1 = type;
+    }
+}
+
+void usbdev_esr(usbdev_t *dev)
+{
+    (void)dev;
+    if (USB->DEVICE.INTFLAG.reg) {
+        if (USB->DEVICE.INTFLAG.bit.EORST) {
+            /* Clear flag */
+            USB->DEVICE.INTFLAG.reg = USB_DEVICE_INTFLAG_EORST;
+            usbdev_ep_init(&endpoints[0]);
+            _ep_enable(&endpoints[0]);
+            usbdev_ep_init(&endpoints[1]);
+            _ep_enable(&endpoints[1]);
+            _usbdev->usbdev.cb(&_usbdev->usbdev, USBDEV_EVENT_RESET);
+        }
+        /* Re-enable the USB IRQ */
+        _enable_irq();
+    }
+}
+
+
+static void _ep_address(usbdev_ep_t *ep)
+{
+    UsbDeviceDescBank *bank = &banks[_get_ep_num(ep->num, ep->dir)];
+
+    bank->ADDR.reg = (uint32_t)ep->buf;
+}
+
+void _ep_set_stall(usbdev_ep_t *ep, usbopt_enable_t enable)
+{
+    UsbDeviceEndpoint *ep_reg = &USB->DEVICE.DeviceEndpoint[ep->num];
+
+    if (ep->dir == USB_EP_DIR_IN) {
+        if (enable) {
+            ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSSET_STALLRQ1;
+        }
+        else {
+            ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSSET_STALLRQ1;
+        }
+    }
+    else {
+        if (enable) {
+            ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSSET_STALLRQ0;
+        }
+        else {
+            ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSSET_STALLRQ0;
+        }
+    }
+}
+
+usbopt_enable_t _ep_get_stall(usbdev_ep_t *ep)
+{
+    usbopt_enable_t res;
+    UsbDeviceEndpoint *ep_reg = &USB->DEVICE.DeviceEndpoint[ep->num];
+
+    if (ep->dir == USB_EP_DIR_IN) {
+        res = ep_reg->EPSTATUSSET.bit.STALLRQ1 
+            ? USBOPT_ENABLE 
+            : USBOPT_DISABLE;
+    }
+    else {
+        res = ep_reg->EPSTATUSSET.bit.STALLRQ0
+            ? USBOPT_ENABLE
+            : USBOPT_DISABLE;
+    }
+    return res;
+
+}
+
+static void _ep_size(usbdev_ep_t *ep)
+{
+    UsbDeviceDescBank *bank = &banks[_get_ep_num(ep->num, ep->dir)];
+    unsigned val = 0x00;
+
+    switch (ep->len) {
+        case 8:
+            val = 0x0;
+            break;
+        case 16:
+            val = 0x1;
+            break;
+        case 32:
+            val = 0x2;
+            break;
+        case 64:
+            val = 0x3;
+            break;
+        case 128:
+            val = 0x4;
+            break;
+        case 256:
+            val = 0x5;
+            break;
+        case 512:
+            val = 0x6;
+            break;
+        case 1023:
+            val = 0x7;
+            break;
+        default:
+            return;
```

Comment:
What about a single line to directly compute this value with a nice comment ? :)

Reply:
Should be possible, but have to think about it a bit, the last case (1023) makes it a bit annoying.

Reply:
I've been thinking a bit about this. I'd rather keep it this way as signalling error back in case of invalid size is easier to do with the switch case statement